cardano-db-sync/src/Cardano/DbSync/OffChain.hs

```diff
@@ -0,0 +1,283 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+
+module Cardano.DbSync.OffChain (
+  insertOffChainPoolResults,
+  insertOffChainVoteResults,
+  loadOffChainPoolWorkQueue,
+  loadOffChainVoteWorkQueue,
+  runFetchOffChainPoolThread,
+  runFetchOffChainVoteThread,
+  fetchOffChainPoolData,
+  fetchOffChainVoteData,
+) where
+
+import Cardano.BM.Trace (Trace, logInfo)
+import qualified Cardano.Db as DB
+import Cardano.DbSync.Api
+import Cardano.DbSync.Api.Types (InsertOptions (..), SyncEnv (..))
+import Cardano.DbSync.OffChain.Http
+import Cardano.DbSync.OffChain.Query
+import Cardano.DbSync.Types
+import Cardano.Prelude
+import Control.Concurrent.Class.MonadSTM.Strict (
+  StrictTBQueue (..),
+  flushTBQueue,
+  isEmptyTBQueue,
+  readTBQueue,
+  writeTBQueue,
+ )
+import Control.Monad.Trans.Control (MonadBaseControl)
+import Data.Time.Clock.POSIX (POSIXTime)
+import qualified Data.Time.Clock.POSIX as Time
+import Database.Persist.Sql (SqlBackend)
+import qualified Network.HTTP.Client as Http
+import Network.HTTP.Client.TLS (tlsManagerSettings)
+
+---------------------------------------------------------------------------------------------------------------------------------
+-- Load OffChain Work Queue
+---------------------------------------------------------------------------------------------------------------------------------
+data LoadOffChainWorkQueue a m = LoadOffChainWorkQueue
+  { lQueue :: StrictTBQueue IO a
+  , lRetryTime :: a -> Retry
+  , lGetData :: MonadIO m => POSIXTime -> Int -> ReaderT SqlBackend m [a]
+  }
+
+loadOffChainPoolWorkQueue ::
+  (MonadBaseControl IO m, MonadIO m) =>
+  Trace IO Text ->
+  StrictTBQueue IO OffChainPoolWorkQueue ->
+  ReaderT SqlBackend m ()
+loadOffChainPoolWorkQueue trce workQueue =
+  loadOffChainWorkQueue
+    trce
+    LoadOffChainWorkQueue
+      { lQueue = workQueue
+      , lRetryTime = oPoolWqRetry
+      , lGetData = getOffChainPoolData
+      }
+
+loadOffChainVoteWorkQueue ::
+  (MonadBaseControl IO m, MonadIO m) =>
+  Trace IO Text ->
+  StrictTBQueue IO OffChainVoteWorkQueue ->
+  ReaderT SqlBackend m ()
+loadOffChainVoteWorkQueue trce workQueue =
+  loadOffChainWorkQueue
+    trce
+    LoadOffChainWorkQueue
+      { lQueue = workQueue
+      , lRetryTime = oVoteWqRetry
+      , lGetData = getOffChainVoteData
+      }
+
+loadOffChainWorkQueue ::
+  (MonadBaseControl IO m, MonadIO m) =>
+  Trace IO Text ->
+  LoadOffChainWorkQueue a m ->
+  ReaderT SqlBackend m ()
+loadOffChainWorkQueue _trce offChainWorkQueue = do
+  whenM (liftIO $ atomically (isEmptyTBQueue (lQueue offChainWorkQueue))) $ do
+    now <- liftIO Time.getPOSIXTime
+    runnablePools <- filter (isRunnable now offChainWorkQueue) <$> lGetData offChainWorkQueue now 100
```

Comment:
```suggestion
    runnableOffChainData <- filter (isRunnable now offChainWorkQueue) <$> lGetData offChainWorkQueue now 100
```

since this is generic